Src/ILGPU/Runtime/ArrayViewExtensions.Generated.tt

```diff
@@ -0,0 +1,800 @@
+                    int elementIndex = stride.ComputeElementIndex(multiDimIndex);
+                    data[<#= string.Join(", ",
+                        Enumerable.Range(0, dimension)
+                        .Select(t => IndexDimensions[t].FieldName)) #>] =
+                        span[elementIndex];
+<#      for (int i = 0; i < dimension; ++i) { #>
+                }
+<#      } #>
+            }
+<#      } #>
+        }
+
+        /// <summary>
+        /// Copies the contents of the <#= dimension #>D view into the given
+        /// <#= dimension #>D array using the default accelerator stream.
+        /// </summary>
+        /// <param name="buffer">The source buffer.</param>
+        /// <param name="data">The target data array.</param>
+        /// <remarks>
+<#      if (needsTranspose) { #>
+        /// CAUTION: this method transposes the data on the CPU.
+<#      } #>
+        /// This method is not supported on accelerators.
+        /// </remarks>
+        [NotInsideKernel]
+        public static void CopyToCPU<T>(
+            this MemoryBuffer<
+                <#= typeName #><T, <#= strideTypeName #>.<#= strideName #>>> buffer,
+            <#= arrayType #> data)
+            where T : unmanaged =>
+            buffer.View.CopyToCPU(data);
+
+        /// <summary>
+        /// Copies the contents of the <#= dimension #>D view into the given
+        /// <#= dimension #>D array using the given accelerator stream.
+        /// </summary>
+        /// <param name="buffer">The source buffer.</param>
+        /// <param name="stream">The used accelerator stream.</param>
+        /// <param name="data">The target data array.</param>
+        /// <remarks>
+<#      if (needsTranspose) { #>
+        /// CAUTION: this method transposes the data on the CPU.
+<#      } #>
+        /// This method is not supported on accelerators.
+        /// </remarks>
+        [NotInsideKernel]
+        public static void CopyToCPU<T>(
+            this MemoryBuffer<
+                <#= typeName #><T, <#= strideTypeName #>.<#= strideName #>>> buffer,
+            AcceleratorStream stream,
+            <#= arrayType #> data)
+            where T : unmanaged =>
+            buffer.View.CopyToCPU(stream, data);
+
+        /// <summary>
+        /// Copies the contents of the <#= dimension #>D array into the given
+        /// <#= dimension #>D view using the default accelerator stream.
+        /// </summary>
+        /// <param name="view">The source view.</param>
+        /// <param name="data">The target data array.</param>
+        /// <remarks>
+<#      if (needsTranspose) { #>
+        /// CAUTION: this method transposes the data on the CPU.
+<#      } #>
+        /// This method is not supported on accelerators.
+        /// </remarks>
+        [NotInsideKernel]
+        public static void CopyFromCPU<T>(
+            this <#= typeName #><T, <#= strideTypeName #>.<#= strideName #>> view,
+            <#= arrayType #> data)
```

Comment:
ISSUE: Should replace the overloads of `CopyFromCPU`/`CopyToCPU` that use `arrayType` with `ReadOnlySpan<T>`.

This will provide support for arrays, as well as spans.

Reply:
What about multidimensional arrays? Should we still emit extension methods taking `Span`s instead?

Reply:
ah, true. did not consider that. it would also not work for jagged arrays.